### PR TITLE
Cosmo param dataclass

### PR DESCRIPTION
**Pull request by nstarman**
### Description

Cosmology Parameters are now [dataclass](https://docs.python.org/3.10/library/dataclasses.html).
Why? b/c dataclasses are nice:

- clean, modern code & syntax
- convenient how it writes the dunder methods
- more introspectable, using tools from the `dataclasses` module.
- work with ``dataclasses``'s functions
- easy to build subclasses.

In this PR:

- Add a compatibility flag for python 3.10
- Refactored Parameter to be a dataclass
    - Due to the nature of dataclases, the private attribute `_registry_validators` had to be factored off of the class.
    - Got to remove the `_init_arguments` because dataclasses have `replace`!
    - ~Changed `fvalidate` from the processed value to the original one. This is a breaking change.~

Future note: in py3.13+ https://peps.python.org/pep-0712/ will allow the ``UnitField`` class to be replaced with the ``converter`` argument to ``field()``. In the meantime, it's a private dataclass field descriptor.

### CODE REVIEW

**Comment by nstarman** [comment]
This PR is ready, except that the docs are failing because they can't find the docstring for some descriptors.
@pllim @mhvk, any suggestions?  I found from poking around on numpydoc that you've both tackled weird sphinx docs failures...

**Comment by nstarman** [comment]
> Probably because you have no docstring for those attributes and Sphinx is trying to auto-generate something but then is choking because of private module?

I used to have docstrings, but removed them in
https://github.com/astropy/astropy/pull/14874/commits/527a0393082c031de72e6ada098ff34c229f72ff because it gives the same error! I think it's something to do with how the attribute is being linked to the class of the attribute?

**Comment by pllim** [comment]
> I think it's something to do with how the generated docstring is being linked to the class of the attribute?

Alas, I am not familiar with this particular error. You can try look in `modeling` and see how they do it.

**Comment by nstarman** [comment]
Thanks! figured it out. Sphinx doesn't like semi-public descriptors. `UnitField` -> `_UnitField` fixes everything.

**Comment by nstarman** [comment]
@WilliamJamieson @mhvk @eerovaher, I got all the checks working!

**Comment by mhvk** [comment]
@nstarman - the cosmology initialization was already hard to follow, but I fear this has moved beyond my python skills -- no idea e.g. why one would need `object.__setattr__` etc. I'll unsubscribe from this one.

**Comment by nstarman** [comment]
> @nstarman - the cosmology initialization was already hard to follow, but I fear this has moved beyond my python skills

Thanks regardless for all the other PR reviews!

>  -- no idea e.g. why one would need `object.__setattr__` etc. I'll unsubscribe from this one.

This is because frozen dataclasses cannot be assigned to post initialization and the only way to do so is using ``object.__setattr__`` (see https://docs.python.org/3.10/library/dataclasses.html#frozen-instances).
We need this in Parameter because dataclass descriptor fields are only introduced in py3.10+ and in py3.9 are just descriptors. Thus we need to manually call them since they are not understood to be fields. Next year, when we bump to 3.10+ initialization will be much simpler.

**Comment by nstarman** [comment]
Ping @eerovaher @WilliamJamieson, if you have the time to review this, I'd appreciate your feedback on this PR. I'm looking to make Cosmology a dataclass and this is a first step.

**Comment by nstarman** [comment]
@eerovaher @WilliamJamieson, I figured out how to prevent any API changes introduced by this PR.

**Comment by nstarman** [comment]
Just squashed that commit into the main set

**Comment by nstarman** [comment]
Ping @eerovaher @WilliamJamieson, if you have the time to review this, I'd appreciate your feedback on this PR. I'm looking to make Cosmology a dataclass and this is a first step. Thanks!

File: astropy/cosmology/parameter/_core.py
Comment (eerovaher) [inline astropy/cosmology/parameter/_core.py:110]: Is this supposed to be a string literal instead of a comment?

File: astropy/cosmology/tests/test_parameter.py
Comment (eerovaher) [inline astropy/cosmology/tests/test_parameter.py:119]:
```python
        assert all_parameter.name == f"_{all_parameter._attr_name}"
```

File: astropy/cosmology/parameter/_core.py
Comment (eerovaher) [inline astropy/cosmology/parameter/_core.py:152]: I don't see why is this change in code is being made. According to the type hint `name` is `str` and the empty string is the only falsy string.

File: astropy/cosmology/parameter/_core.py
Comment (eerovaher) [inline astropy/cosmology/parameter/_core.py]: Isn't this comment now outdated?

File: docs/whatsnew/6.0.rst
Comment (eerovaher) [inline docs/whatsnew/6.0.rst:114]:
```suggestion
This means that the :mod:`dataclasses` machinery can be used to work with
```

File: docs/changes/cosmology/14874.api.rst
Comment (eerovaher) [inline docs/changes/cosmology/14874.api.rst:1]:
```suggestion
Cosmology ``Parameter`` is now a ``dataclass``, and can work with all of Python's dataclasses
```

**Review by eerovaher — COMMENTED** [review]
This is a good idea because utilizing `dataclass` can make the code more familiar to anyone that knows about that part of the Python standard library, and it also simplifies the implementation of the desired functionality. Note that although the pull request adds more lines then it removes, the added What's New and changelog entries together with compatibility code for Python < 3.10 all contribute to the line count.

I do have a few questions and comments about some of the details.

File: astropy/cosmology/parameter/_core.py
Comment (nstarman) [reply astropy/cosmology/parameter/_core.py:24]: in no small part, yes. Python 3.10's descriptors-as-fields mechanic is great, but should not be touched unless one really knows what they are doing. Freezing this dataclass helps.
More generally, immutable structures are easier as they are more stable, enabling e.g. caching without a sophisticated updating mechanism that is required if fields are mutable. dataclasses make it easy to create immutable structures in Python, which is another excellent reason to use dataclasses!

File: astropy/cosmology/parameter/_core.py
Comment (nstarman) [reply astropy/cosmology/parameter/_core.py:27]: > Are you sure that `u.Unit` is the correct type here and not `u.UnitBase`

The `__setattr__` always wraps the value as `u.Unit(value)`, so `u.Unit` should be the correct type here.
(a strange emphasis on should, since `u.Unit` is strange).

> what would be the downside of replacing None with `u.dimensionless_unscaled`

If the value doesn't have a unit, as in it isn't a `Quantity`, which is distinct from `u.dimensionless_unscaled`.

> The upside would be that the return type wouldn't be a union type.

Agreed, but see above. As a note, part of my plan for Cosmology is to make more fields, e.g. `Om0`, have units, even if the unit is `u.dimensionless_unscaled`.

File: astropy/cosmology/parameter/_core.py
Comment (nstarman) [reply astropy/cosmology/parameter/_core.py:107]: As in  `unit: u.Unit = field(default_factory=_UnitField)`?

That might work. I'll need to check that `_UnitField()` gets property attached as a descriptor and calls `__set_name__`. It's not functionally different than the current setup (https://docs.python.org/3/library/dataclasses.html#descriptor-typed-fields). Is there a reason you are suggesting this alternative?

File: astropy/cosmology/parameter/_core.py
Comment (nstarman) [reply astropy/cosmology/parameter/_core.py:132]: They are subtly different (https://docs.python.org/3/c-api/object.html?highlight=__class__#c.PyObject_IsInstance). A good question is whether we want to allow a user to muck around with the constructor in this way. `type(x)` is safer, `x.__class__` is more flexible.

File: astropy/cosmology/tests/test_parameter.py
Comment (nstarman) [reply astropy/cosmology/tests/test_parameter.py:131]: Agreed. I'll remove this test!

File: astropy/cosmology/parameter/_core.py
Comment (nstarman) [reply astropy/cosmology/parameter/_core.py:110]: Sphinx can (/should) pick up on comments directly after fields and display it in the docs.
In my experience the mechanism is flaky.

File: astropy/cosmology/tests/test_parameter.py
Comment (nstarman) [reply astropy/cosmology/tests/test_parameter.py:119]: Other way around?

File: astropy/cosmology/parameter/_core.py
Comment (nstarman) [reply astropy/cosmology/parameter/_core.py:152]: Let me fix that. I see that the ``__set_name__`` constructor is being called with ``name=None`` in one location.

File: astropy/cosmology/parameter/_core.py
Comment (nstarman) [reply astropy/cosmology/parameter/_core.py]: No, Python will call ``__set_name__`` a second time, if the instance is correctly attached as descriptor (i.e. constructed as another class's field, not attached afterwords). This first time is to ensure that all attributes exist, in case the class is not constructed as a descriptor.

Comment (nstarman) [reply astropy/cosmology/parameter/_core.py]: I'll update the comment to make it more clear.

Comment (nstarman) [reply astropy/cosmology/parameter/_core.py]: Done.

File: astropy/cosmology/parameter/_core.py
Comment (nstarman) [reply astropy/cosmology/parameter/_core.py]: Done.

File: astropy/cosmology/tests/test_parameter.py
Comment (nstarman) [reply astropy/cosmology/tests/test_parameter.py]: Done.

File: astropy/cosmology/parameter/_core.py
Comment (nstarman) [reply astropy/cosmology/parameter/_core.py]: Done.

File: astropy/cosmology/parameter/_core.py
Comment (nstarman) [reply astropy/cosmology/parameter/_core.py:107]: Looks like it works, but given the [example in the docs](https://docs.python.org/3/library/dataclasses.html#descriptor-typed-fields), there's no need to wrap `_UnitField` within a `field`.

File: astropy/cosmology/parameter/_core.py
Comment (eerovaher) [reply astropy/cosmology/parameter/_core.py:29]: `u.Unit` is indeed strange:
```python
>>> from astropy import units as u
>>> isinstance(u.m, u.Unit)
False
>>> type(u.m).__mro__
(<class 'astropy.units.core.IrreducibleUnit'>, <class 'astropy.units.core.NamedUnit'>, <class 'astropy.units.core.UnitBase'>, <class 'object'>)
```

I would expect that all cosmology parameters could be represented as `Quantity` instances, even if some of them would require `u.dimensionless_unscaled` as the unit, but if that is not the case currently then it is indeed better to leave it as it is for now.

File: astropy/cosmology/parameter/_core.py
Comment (eerovaher) [reply astropy/cosmology/parameter/_core.py:107]: Using `default_factory` would get rid of the `noqa` instruction.

File: astropy/cosmology/parameter/_core.py
Comment (eerovaher) [reply astropy/cosmology/parameter/_core.py:110]: Sphinx seems to indeed pick up many of these ([link to `Parameter` docs](https://astropy--14874.org.readthedocs.build/en/14874/api/astropy.cosmology.Parameter.html#astropy.cosmology.Parameter)), but I don't see any documentation for `equivalencies`. Is that how it is supposed to be?

File: astropy/cosmology/parameter/_core.py
Comment (nstarman) [reply astropy/cosmology/parameter/_core.py:107]: True, however I think https://github.com/astral-sh/ruff/pull/5537 should be coming down the release pipe, and will likewise remove the `noqa`.

File: astropy/cosmology/parameter/_core.py
Comment (nstarman) [reply astropy/cosmology/parameter/_core.py:110]: If it's not picking it up, I'll remove the comment. I think the field is largely self explanatory, so not work the effort wrangling Sphinx.

File: astropy/cosmology/parameter/_core.py
Comment (nstarman) [reply astropy/cosmology/parameter/_core.py:132]: But I don't anticipate users messing with this, so probably worth changing to `type(x)`

File: astropy/cosmology/parameter/_core.py
Comment (nstarman) [reply astropy/cosmology/parameter/_core.py:107]: Good to keep as-is for now?

File: astropy/cosmology/parameter/_core.py
Comment (nstarman) [reply astropy/cosmology/parameter/_core.py]: Done!

File: astropy/cosmology/parameter/_core.py
Comment (nstarman) [reply astropy/cosmology/parameter/_core.py]: Done!

File: astropy/cosmology/tests/test_parameter.py
Comment (nstarman) [reply astropy/cosmology/tests/test_parameter.py]: Done!

File: astropy/cosmology/parameter/_core.py
Comment (eerovaher) [reply astropy/cosmology/parameter/_core.py:106]: Although I recognize the need for some `noqa` instructions, the fewer of them we have the better. But in this case it does look like the problem is not in our code but in the Ruff implementation of the rule, so it is indeed better to leave the code as is. We just need to make sure that we actually remove the `noqa` when we update Ruff.

File: astropy/cosmology/parameter/_core.py
Comment (eerovaher) [reply astropy/cosmology/parameter/_core.py]: Sphinx did recognize the docstrings for what is listed under the Attributes Summary and those docstrings were nice to have. But for some reason `equivalencies` is not listed among the attributes at all.

Comment (nstarman) [reply astropy/cosmology/parameter/_core.py]: Strange, when I build the docs myself it works great. Trying again with RTD.

CleanShot 2023-07-28 at 16 39 57@2x

Comment (nstarman) [reply astropy/cosmology/parameter/_core.py]: It works on RTD now!

**Comment by nstarman** [comment]
RTD is rendering the attributes. https://astropy--14874.org.readthedocs.build/en/14874/api/astropy.cosmology.Parameter.html#astropy.cosmology.Parameter

File: docs/whatsnew/6.0.rst
Comment (eerovaher) [inline docs/whatsnew/6.0.rst:132]:
```suggestion
Parameter as a :func:`~dataclasses.dataclass`
```

**Review by eerovaher — CHANGES_REQUESTED** [review]
I just have two small remarks left. But perhaps it is best to wait until #15065 or #15096 have been merged so that the CI tests wouldn't fail because of the new `numpy` scalar representations.

File: astropy/cosmology/parameter/_core.py
Comment (nstarman) [reply astropy/cosmology/parameter/_core.py:60]: I wholeheartedly agree on both points! However, I think this PR manages not to change any API, which is why I didn't consolidate nor cleanup the error messages.
I'll open an Issue to do this as a followup.

File: docs/whatsnew/6.0.rst
Comment (eerovaher) [inline docs/whatsnew/6.0.rst:133]: The title underline being too short is breaking the documentation build, so while we are editing this anyways we might as well add another hyperlink.
```suggestion
:class:`~astropy.cosmology.Parameter` as a :func:`~dataclasses.dataclass`
-------------------------------------------------------------------------
```

**Review by eerovaher — CHANGES_REQUESTED** [review]

File: astropy/cosmology/parameter/_core.py
Comment (eerovaher) [reply astropy/cosmology/parameter/_core.py:60]: I see you have already opened an issue for this.

File: docs/whatsnew/6.0.rst
Comment (nstarman) [reply docs/whatsnew/6.0.rst]: Done!

**Review by eerovaher — CHANGES_REQUESTED** [review]
The diff on GitHub looks good enough to merge, but when I looked at the [What's New page on RTD](https://astropy--14874.org.readthedocs.build/en/14874/whatsnew/6.0.html) I noticed that the new entry is not under the right section. Other than that this looks good to go.

**Comment by nstarman** [comment]
The pains of rebasing! Thanks for noticing.

**Review by eerovaher — APPROVED** [review]

**Comment by nstarman** [comment]
Thank you @eerovaher for the many valuable reviews!

**Pull request merged**
